DRY up some common remove utils from mutable.Queue to ArrayDeque

pathikrit opened the pull request:
Adds removeFirst(predicate), removeAll(predicate) to ArrayDeque
Adds dequeWhile(predicate) to mutable.Queue

lrytz commented on src/library/scala/collection/mutable/ArrayDeque.scala:
```diff
@@ -51,22 +51,22 @@ class ArrayDeque[A] protected (
   def this(initialSize: Int = ArrayDeque.DefaultInitialSize) = this(ArrayDeque.alloc(initialSize), start = 0, end = 0)
 
-  def apply(idx: Int) = {
+  override def apply(idx: Int) = {
```
we ususally don't write `override` for implementations of abstract methods

pathikrit replied:
Isn't putting `override` a good idea? If the parent removes the abstract method, it would be a compile error instead of being left with an orphan method?

lrytz replied:
It's a matter of taste, I personally see it like https://stackoverflow.com/a/5643284/248998. In this repo we use `override` when actually overriding.

pathikrit replied:
This convention put a lot of burden on the author of the parent - should she decide to provide a default implementation one day, she would have to hunt down all the subclasses and remove the override keyword. But, as you said, its a matter of taste - I don't have a strong opinion here since my contribution to this repository is tiny :)

lrytz commented on src/library/scala/collection/mutable/ArrayDeque.scala:
```diff
@@ -83,7 +83,7 @@ class ArrayDeque[A] protected (
   }
 
-  override def prependAll(elems: IterableOnce[A]): this.type = {
+  override def prependAll(elems: IterableOnce[A]) = {
```
We should keep the explicit return types (also in other parts of this PR)

pathikrit replied:
Isn't the return type redundant (and error-free) when used alongside the override keyword? I am fine either way - I just saw some other places in the code which did not have return type for overrides so I chose to follow one convention. But, I can switch.

lrytz replied:
You're right that the return type is "inherited", but as far as I know we (try to) put the types on public members also in this case.

pathikrit replied:
Ok will do!

lrytz reviewed (commented):
`removeFirst` and `removeAll` on ArrayDeque are new methods. Looking at ArrayDeque, there are other such methods, e.g., `removeHeadOption`, `removeHead`, `removeLastOption`, `removeLast`, `removeHeadWhile`, `removeLastWhile`. Would some of them make sense higher up in the hieararchy?

I think it might also be confusing for users of `Queue` to have both `dequeueFirst` and `removeFirst` available, both doing the same thing.

pathikrit commented:
I think Queue and Stack should not extend ArrayDeque but have it as a private member (composition over inheritance)

julienrf commented:
> I think Queue and Stack should not extend ArrayDeque but have it as a private member (composition over inheritance)

That could solve a few issues with their current implementation, indeed. You can give that a try (in a separate PR?) if you want.

szeiger reviewed (commented):
I like moving the methods up into `ArrayDeque`, this could be useful. I'd prefer to go without the "clean-ups" (removing return types, adding `override`).

szeiger commented:
@pathikrit Do you think you'll find the time to finish this? We'd like to drain the queue for M5 until the end of this week.

pathikrit commented:
@szeiger : Ok let me jump on this now

pathikrit committed:
DRY up some common remove utils from mutable.Queue to ArrayDeque
Adds removeFirst(predicate), removeAll(predicate) to ArrayDeque
Adds dequeWhile(predicate) to mutable.Queue

pathikrit commented:
@szeiger : Addressed feedback.

SethTisue commented:
thank you Pathikrit!

pathikrit commented:
No problem - next up - we should do this: https://github.com/scala/scala/pull/6762#issuecomment-396244365 
I can send up a PR for that
